media_driver/agnostic/common/codec/hal/codechal_vdenc_hevc.cpp

```diff
@@ -796,14 +799,18 @@ MOS_STATUS CodechalVdencHevcState::SetupMbQpStreamIn(PMOS_RESOURCE streamIn)
             //            uint32_t YOffset = (h % 2) * 2;
             //            uint32_t XOffset = 2 * (w/2 * 2) + w % 2;
 
-            int32_t ForceQp = pInputData[(h/2) * m_encodeParams.psMbQpDataSurface->dwPitch + (w/2)];
             //            (pData + (Offset + XOffset + YOffset))->DW7.QpEnable = 0xf;
             //            (pData + (Offset + XOffset + YOffset))->DW14.ForceQp_0 = ForceQp;
             //            (pData + (Offset + XOffset + YOffset))->DW14.ForceQp_1 = ForceQp;
             //            (pData + (Offset + XOffset + YOffset))->DW14.ForceQp_2 = ForceQp;
             //            (pData + (Offset + XOffset + YOffset))->DW14.ForceQp_3 = ForceQp;
             streaminDataParams.setQpRoiCtrl = true;
-            streaminDataParams.forceQp = (int8_t)ForceQp;
+
+            streaminDataParams.forceQp[0] = (int8_t) ( pInputData[(h * 2)     * m_encodeParams.psMbQpDataSurface->dwPitch + (w * 2)]);
+            streaminDataParams.forceQp[1] = (int8_t) ( pInputData[(h * 2)     * m_encodeParams.psMbQpDataSurface->dwPitch + (w * 2 + 1)]);
```

Comment:
Hi Xiaoxia, reading pInputData four times in this loop will significantly increase the CPU utilization and impact the performance.

Reply:
> Hi Xiaoxia, reading pInputData four times in this loop will significantly increase the CPU utilization and impact the performance.

Yes, it will. But for 16x16 CU QP, it has to read four QP values for each 16x16 cu block from the QP map table. Do you have suggestion to improve it?

Reply:
Copy the buffer to system memory before the loop can get lower CPU utilization and better performance.

Reply:
Thank you for your suggestion, Tony. I have updated the code as you said. Please check the change in below new patch.

Reply:
Thanks for the update, Xiaoxia. The change looks good to me.